### PR TITLE
Backend image rendering as plugin

### PR DESCRIPTION
**Pull request by torkelo**
The most stable and supported way to automate headless chrome is using the official Google node framework [puppeteer](https://github.com/GoogleChrome/puppeteer). This requires that rendering is done from a node process. A way to simplify the management of that process & distribution would be to leverage the backend plugin model where hashicorps plugin framework can manage some of the process management and grpc will handle the communication. 

* [x] Introduce new plugin type `renderer`
* [x] Node app that can both be started as grpc Grafana backend plugin or as http server https://github.com/grafana/grafana-image-renderer

The idea is that the node app can also act as a standalone http server. In this case you will have to configure Grafana to use it. The use case here would be in hosted scenarios where you want to share rendering resources. 

In plugin mode, you will have to git clone the image renderer plugin repo into the plugins folder and run "yarn install & npm run build". 

* [x] Pass render key
* [x] Add timeout handling
* [x] Wait for rendering to complete 
* [x] Timezone handling
* [x] Support grpc plugin mode
* [x] Support HTTP mode
* [x] Fallback to current phantomjs code

One way forward would to default to phantomjs but if the new renderer plugin is found then that is used instead. This way we can merge this before solving the distribution problem (outlined below). 

Open questions:
* [ ] Distribution of the plugin, since it requires node & npm is currently the one installing the platform specific version of chromium. Hard to package a multiplatform version of it, will require node for all platforms & chromium for all platforms checked into repo. So not sure that is the solution. For docker it simpler we can just install it during Docker build. 

thoughts @bergquist @daniellee @DanCech

### CODE REVIEW

**Comment by mauve** [comment]
Hi, Shall I abandon this PR? https://github.com/grafana/grafana/pull/10751

**Comment by torkelo** [comment]
Yes , i tried it but had so many stability issues and crashes with chromedp that I found it not good option for this.

**Comment by torkelo** [comment]
This now fallbacks to the existing method of using phantomjs. Everything is handled via the new RenderService.  The phantomjs command execution code could be quite nicely simplified as well using new golang library methods for executing commands with context timeouts. The render call now correctly has a context and everything is canceled & phantomjs is shutdown if http request is aborted  (or alerting rule eval context is canceled if rendering is done in this context). 

So this is actually mergable (if tests pass). The plugin mode still needs a little more work but not much, and need to add support for http mode. 

Think the way we move forward is to release this in v5.2 as is but with optional plugin & http mode support. that way we can start testing it and start testing HG with external render server. When that is stable we can talk about how to remove phantomjs support in v5.3 and how to make the plugin easy to install for all non docker users.

**Comment by torkelo** [comment]
I think this is mergable now. 

Tested with phantomjs (works as before), and if you install the plugin it will be used instead. You can now optionally also set 

```ini
[rendering]
server_url = http://my.grafana-image-renderer.com
```

The https://github.com/grafana/grafana-image-renderer http server mode needs a lot more work to be more scalable & some limits so a single instance cannot do 1000 requests and run it out off memory. 

## Backend plugins refactoring

Since this introduced another backend plugin it also caused some duplication in the code that manages the start & keep alive of backend plugins. I have started some work on plugins package refactoring that will make new backend plugins a lot easier to add so they can all share the same startup & keep alive code. But this will be in a another PR. 

Here is taste of it. 

The idea is to replace this 
```go
PluginTypes = map[string]interface{}{
		"panel":      PanelPlugin{},
		"datasource": DataSourcePlugin{},
		"app":        AppPlugin{},
		"renderer":   RendererPlugin{},
	}
```

With more detailed descriptor type (and have plugin types self-register).

```go
type RendererPlugin struct {
	PluginBase
	Executable string `json:"executable,omitempty"`
}

func init() {
	RegisterPluginType(&PluginTypeDescriptor{
		Id:               "renderer",
		PluginModel:      RendererPlugin{},
		ProtocolVersion:  1,
		ServiceInterface: grpcProxy.RendererPluginImpl{},
	})
}

```

File: pkg/services/rendering/http_mode.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	// save response to file
+	defer resp.Body.Close()
+	out, err := os.Create(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer out.Close()
+	io.Copy(out, resp.Body)
```
Comment (bergquist) [inline pkg/services/rendering/http_mode.go]: shouldn't we handle this error?

Comment (torkelo) [reply pkg/services/rendering/http_mode.go]: yes, your right. The error returned by io.Copy should be handled, missed that

**Pull request merged**